exchanges/exchange.go

```diff
@@ -549,94 +553,148 @@ func (b *Base) SetPairs(pairs currency.Pairs, assetType asset.Item, enabled bool
-	if force || len(newPairs) > 0 || len(removedPairs) > 0 {
-		if force {
+	if force {
+		log.Debugf(log.ExchangeSys,
+			"%s forced update of %s [%v] pairs.",
+			b.Name,
+			updateType,
+			strings.ToUpper(assetType.String()))
+	} else {
+		if len(diff.New) > 0 {
 			log.Debugf(log.ExchangeSys,
-				"%s forced update of %s [%v] pairs.",
+				"%s Updating %s pairs [%v] - Added: %s.\n",
 				b.Name,
 				updateType,
-				strings.ToUpper(assetType.String()))
-		} else {
-			if len(newPairs) > 0 {
-				log.Debugf(log.ExchangeSys,
-					"%s Updating %s pairs [%v] - Added: %s.\n",
-					b.Name,
-					updateType,
-					strings.ToUpper(assetType.String()),
-					newPairs)
-			}
-			if len(removedPairs) > 0 {
-				log.Debugf(log.ExchangeSys,
-					"%s Updating %s pairs [%v] - Removed: %s.\n",
-					b.Name,
-					updateType,
-					strings.ToUpper(assetType.String()),
-					removedPairs)
-			}
+				strings.ToUpper(assetType.String()),
+				diff.New)
 		}
+		if len(diff.Remove) > 0 {
+			log.Debugf(log.ExchangeSys,
+				"%s Updating %s pairs [%v] - Removed: %s.\n",
+				b.Name,
+				updateType,
+				strings.ToUpper(assetType.String()),
+				diff.Remove)
+		}
+	}
 
-		b.Config.CurrencyPairs.StorePairs(assetType, products, enabled)
-		b.CurrencyPairs.StorePairs(assetType, products, enabled)
+	b.Config.CurrencyPairs.StorePairs(assetType, incomingPairs, enabled)
+	b.CurrencyPairs.StorePairs(assetType, incomingPairs, enabled)
 
-		if !enabled {
-			// If available pairs are changed we will remove currency pair items
-			// that are still included in the enabled pairs list.
-			enabledPairs, err := b.CurrencyPairs.GetPairs(assetType, true)
-			if err == nil {
-				return nil
-			}
-			_, remove := enabledPairs.FindDifferences(products)
-			for i := range remove {
-				enabledPairs = enabledPairs.Remove(remove[i])
-			}
+	if !enabled {
+		return b.inspectEnabledPairs(incomingPairs, assetType, pFmt)
+	}
 
-			if len(remove) > 0 {
-				log.Debugf(log.ExchangeSys,
-					"%s Checked and updated enabled pairs [%v] - Removed: %s.\n",
-					b.Name,
-					strings.ToUpper(assetType.String()),
-					remove)
-				b.Config.CurrencyPairs.StorePairs(assetType, enabledPairs, true)
-				b.CurrencyPairs.StorePairs(assetType, enabledPairs, true)
+	return nil
+}
+
+// inspectEnabledPairs checks for differences after an available pairs
+// adjustment which will remove currency pairs that have been disabled by an
+// exchange or adjust the entire list of enabled pairs if there is a formatting
+// change. It will also capture unintentional client inputs e.g. a client can
+// enter `linkusd` via config and that might be unintentionally be formatted too
+// `lin-kusd` which can cause system discrepencies for fetching, saving and
+// duplication.
+func (b *Base) inspectEnabledPairs(availPairs currency.Pairs, assetType asset.Item, pFmt currency.PairFormat) error {
+	enabledPairs, err := b.CurrencyPairs.GetPairs(assetType, true)
+	if err == nil && !enabledPairs.FormatDifference(pFmt) {
```

Comment:
The only error that is returned is that it isn't contained in the available pairs list and should be removed or changed. I have added a specific exported error so if we have any changes in the future it will return an error correctly.